ResultSet#next() calls past last element throw IOException

```
java.sql.SQLException: java.io.IOException: Attempted read on closed stream.
	at ru.yandex.clickhouse.response.ClickHouseResultSet.hasNext(ClickHouseResultSet.java:132)
	at ru.yandex.clickhouse.response.ClickHouseResultSet.next(ClickHouseResultSet.java:139)
	at ru.yandex.clickhouse.integration.ClickHouseStatementImplTest.testSingleColumnResultSetMultipleHasNextPastLast(ClickHouseStatementImplTest.java:78)
Caused by: java.io.IOException: Attempted read on closed stream.
	at org.apache.http.conn.EofSensorInputStream.isReadAllowed(EofSensorInputStream.java:109)
	at org.apache.http.conn.EofSensorInputStream.read(EofSensorInputStream.java:118)
	at ru.yandex.clickhouse.response.ClickHouseLZ4Stream.readNextBlock(ClickHouseLZ4Stream.java:82)
	at ru.yandex.clickhouse.response.ClickHouseLZ4Stream.checkNext(ClickHouseLZ4Stream.java:74)
	at ru.yandex.clickhouse.response.ClickHouseLZ4Stream.read(ClickHouseLZ4Stream.java:50)
	at ru.yandex.clickhouse.response.StreamSplitter.readFromStream(StreamSplitter.java:88)
	at ru.yandex.clickhouse.response.StreamSplitter.next(StreamSplitter.java:50)
	at ru.yandex.clickhouse.response.ClickHouseResultSet.hasNext(ClickHouseResultSet.java:115)
	... 35 more
```

--- a/src/main/java/ru/yandex/clickhouse/response/ClickHouseResultSet.java
+++ b/src/main/java/ru/yandex/clickhouse/response/ClickHouseResultSet.java
@@ -60,6 +60,8 @@ public class ClickHouseResultSet extends AbstractResultSet {
 
     private boolean usesWithTotals;
 
+    private boolean lastReached = false;
+
     public ClickHouseResultSet(InputStream is, int bufferSize, String db, String table, boolean usesWithTotals, ClickHouseStatement statement, TimeZone timezone, ClickHouseProperties properties) throws IOException {
         this.db = db;
         this.table = table;
@@ -108,19 +110,21 @@ public class ClickHouseResultSet extends AbstractResultSet {
     }
 
     public boolean hasNext() throws SQLException {
-        if (nextLine == null) {
+        if (nextLine == null && !lastReached) {
             try {
                 nextLine = bis.next();
 
                 if (nextLine == null || (maxRows != 0 && rowNumber >= maxRows) || (usesWithTotals && nextLine.length() == 0)) {
                     if (usesWithTotals) {
-                        if(onTheSeparatorRow()) {
+                        if (onTheSeparatorRow()) {
                             totalLine = bis.next();
                             bis.close();
+                            lastReached = true;
                             nextLine = null;
                         } // otherwise do not close the stream, it is single column or invalid result set case
                     } else {
                         bis.close();
+                        lastReached = true;
                         nextLine = null;
                     }
                 }

--- a/src/main/java/ru/yandex/clickhouse/response/ClickHouseResultSet.java
+++ b/src/main/java/ru/yandex/clickhouse/response/ClickHouseResultSet.java
@@ -121,14 +121,10 @@ public class ClickHouseResultSet extends AbstractResultSet {
                     if (usesWithTotals) {
                         if (onTheSeparatorRow()) {
                             totalLine = bis.next();
-                            bis.close();
-                            lastReached = true;
-                            nextLine = null;
+                            endOfStream();
                         } // otherwise do not close the stream, it is single column or invalid result set case
                     } else {
-                        bis.close();
-                        lastReached = true;
-                        nextLine = null;
+                        endOfStream();
                     }
                 }
             } catch (IOException e) {
@@ -137,6 +133,13 @@ public class ClickHouseResultSet extends AbstractResultSet {
         }
         return nextLine != null;
     }
+
+    private void endOfStream() throws IOException {
+        bis.close();
+        lastReached = true;
+        nextLine = null;
+    }
+
     @Override
     public boolean next() throws SQLException {
         if (hasNext()) {

--- a/src/main/java/ru/yandex/clickhouse/response/ClickHouseResultSet.java
+++ b/src/main/java/ru/yandex/clickhouse/response/ClickHouseResultSet.java
@@ -60,6 +60,9 @@ public class ClickHouseResultSet extends AbstractResultSet {
 
     private boolean usesWithTotals;
 
+    // NOTE this can't be used for `isLast` impl because
+    // it does not do prefetch. It is effectively a witness
+    // to the fact that rs.next() returned false.
     private boolean lastReached = false;
 
     public ClickHouseResultSet(InputStream is, int bufferSize, String db, String table, boolean usesWithTotals, ClickHouseStatement statement, TimeZone timezone, ClickHouseProperties properties) throws IOException {

--- a/src/test/java/ru/yandex/clickhouse/integration/ClickHouseStatementImplTest.java
+++ b/src/test/java/ru/yandex/clickhouse/integration/ClickHouseStatementImplTest.java
@@ -64,6 +64,24 @@ public class ClickHouseStatementImplTest {
     }
 
     @Test
+    public void readsPastLastAreSafe() throws SQLException {
+        ResultSet rs = connection.createStatement().executeQuery("select c from (\n" +
+                "    select 'a' as c, 1 as rn\n" +
+                "    UNION ALL select 'b' as c, 2 as rn\n" +
+                "    UNION ALL select '' as c, 3 as rn\n" +
+                "    UNION ALL select 'd' as c, 4 as rn\n" +
+                " ) order by rn");
+        StringBuffer sb = new StringBuffer();
+        while (rs.next()) {
+            sb.append(rs.getString("c")).append("\n");
+        }
+        Assert.assertFalse(rs.next());
+        Assert.assertFalse(rs.next());
+        Assert.assertFalse(rs.next());
+        Assert.assertEquals(sb.toString(), "a\nb\n\nd\n");
+    }
+
+    @Test
     public void testSelectUInt32() throws SQLException {
         Statement stmt = connection.createStatement();
         ResultSet rs = stmt.executeQuery("select toUInt32(10), toUInt32(-10)");

